### PR TITLE
Avoid optional framework imports in recipe CLI metadata

### PR DESCRIPTION
**Pull request by chesterxgchen**
## Summary
- serve `nvflare recipe list` and `nvflare recipe show` from a packaged, generated metadata catalog
- discover recipe classes through static inheritance analysis and extract constructor metadata with AST only when regenerating the catalog
- add a freshness check so new recipes and constructor changes cannot leave the catalog stale
- preserve the existing 25-recipe list output while preventing unrelated optional frameworks from loading

## Motivation
`nvflare recipe list` without a framework filter previously discovered every recipe package dynamically, causing the XGBoost recipe package and its native XGBoost library to be loaded. `nvflare recipe show fedavg-pt` used the same unfiltered discovery path, so it could also fail on macOS when `libomp` was unavailable even though the requested recipe was PyTorch-only. Recipe metadata queries should not require optional framework packages or native runtimes.

## Validation
- `python -m pytest tests/unit_test/tool/recipe_cli_test.py tests/unit_test/tool/agent/agent_json_contract_test.py tests/unit_test/tool/agent_skill_checks/seed_skills_test.py -q` (81 passed)
- `python -m pytest tests/unit_test/tool -q` (2,777 passed, 39 skipped; one sandbox-denied Unix-socket test passed when rerun with socket permission)
- full CLI startup import guard: both `recipe list` and `recipe show fedavg-pt` succeeded while XGBoost imports were rejected
- `python -m nvflare.tool.recipe.generate_recipe_catalog --check`
- `./runtest.sh -s`
- generated catalog confirmed present in installed package metadata

### CODE REVIEW

File: nvflare/tool/recipe/recipe_cli.py [hunk 1]
```diff
@@ -299,96 +299,79 @@ def _module_source_path(module_name: str):
...
+    candidate_names = set()
+    while True:
+        discovered_names = {
+            node.name
+            for node in class_nodes
+            if any(
+                _ast_base_name(base).endswith("Recipe") or _ast_base_name(base) in candidate_names
+                for base in node.bases
+            )
+        }
```
[the remark this answers is not in the document]
Comment (chesterxgchen) [reply nvflare/tool/recipe/recipe_cli.py]: Thanks. `UnifiedFedAvgRecipe` itself already matched the suffix heuristic, but an imported alias such as `FedAvgRecipe as UnifiedBase` could hide the recipe base. Commit b0ed8f42f resolves imported symbols before applying the heuristic and adds regression coverage for that alias pattern.

**Comment by chesterxgchen** [comment]
Addressed the latest review findings in b0ed8f42f:

- Missing, unreadable, or schema-invalid generated catalogs now produce a structured `INTERNAL_ERROR` and exit 5 instead of a Python traceback. Schema validation covers malformed top-level payloads and malformed recipe entries.
- Catalog generation now statically preserves explicit recipe class metadata (framework support, optional dependencies, heterogeneity, privacy compatibility, notes, and template references), so those attributes are no longer ignored. Runtime list/show still read JSON only and do not import optional frameworks.
- Imported recipe base aliases are resolved during static discovery, with regression coverage.

Validation: 73 focused tests passed, catalog freshness passed, `git diff --check` passed, and `./runtest.sh -s` passed.

**Comment by chesterxgchen** [comment]
Addressed the two follow-up review findings in 1a00a3516:

- Static metadata now goes through JSON-safe normalization before entering the generated catalog. Set values are converted to deterministically sorted lists, including the pre-existing string-list normalization paths such as `recipe_privacy`, so catalog rendering and freshness checks remain reproducible.
- Catalog loading now validates required fields and their expected types for both summary and detail objects. It also validates string-list contents, parameter objects, and matching recipe identity fields, so incomplete entries are rejected as structured `INTERNAL_ERROR` responses before `recipe list` or `recipe show` dereferences them.

Regression validation: 75 focused tests passed, catalog freshness passed, `git diff --check` passed, and `./runtest.sh -s` passed.

**Comment by chesterxgchen** [comment]
Addressed the parameter-contract review finding in ca71157b1. Every generated-catalog parameter must now contain `name` (string), `type` (string or null), `required` (boolean), `default` (present; any JSON value), and `kind` (string). A catalog containing `parameters: [{}]` is rejected through the structured `INTERNAL_ERROR` path before `recipe show --format json` can return malformed metadata. Regression coverage also exercises every invalid field type and a missing `default`.

Validation: 76 focused tests passed; catalog freshness, `git diff --check`, and `./runtest.sh -s` passed.

**Comment by chesterxgchen** [comment]
Addressed the second-pass findings in e1b7f6f6a:

1. Static class-attribute extraction now honors alias priority, records the last assignment to each alias, and continues past non-literal aliases to later explicit literal metadata.
2. Removed summary-to-detail reconstruction and the live-class metadata path. Detail JSON is generated once from the full static discovery entry; callers that need details now load the packaged detail section directly.
3. Replaced the class-name suffix heuristic with recursive static inheritance resolution to `nvflare.recipe.spec.Recipe`, including relative imports, imported aliases, re-exports, and cycle guards. This rejects unrelated `*Recipe` classes and recognizes genuine recipes whose class/base aliases lack that suffix.
4. Static docstring and constructor lookup now follow recipe ancestry, matching inherited behavior without importing recipe packages.
5. Documented specs now override only explicitly declared keys, so omitted fields no longer erase discovered metadata; explicit `None` remains meaningful.
6. Detail metadata uses one AST-dictionary lookup path.
7. Attribute aliases are defined once and shared by extraction and lookup.

I intentionally retained validation of both summary and detail sections when loading the 25-entry artifact. The cost is negligible, and whole-artifact validation ensures corruption fails consistently instead of allowing one command view to mask a broken packaged catalog.

Validation: 79 focused tests passed; the broader tool suite produced 2,777 passed and 39 skipped, with its only sandbox-denied Unix-socket test passing outside the sandbox. Catalog freshness, `git diff --check`, and `./runtest.sh -s` passed. The generated catalog remains byte-for-byte unchanged.

**Comment by chesterxgchen** [comment]
Addressed review #166 in `fb85965c2`:

- added static C3 MRO calculation and reused it for recipe ancestry, inherited docstrings, inherited constructors, and metadata
- apply inherited metadata in reverse MRO order so `Child(Left, Right)` resolves conflicting attributes from `Left`, matching Python
- added an explicit missing-value sentinel so `recipe_notes = None` shadows inherited metadata
- added regression coverage for both multiple-inheritance precedence and explicit clearing

Validation: focused recipe/contract/seed-skill suite (80 passed), catalog freshness check, `git diff --check`, `./runtest.sh -s`, and pre-push agent-skill lint all passed.

**Comment by chesterxgchen** [comment]
Addressed review #167 in `f2711f4c3`:

- resolve algorithm, aggregation, state-exchange, and privacy attributes through the static C3 MRO
- resolve detail metadata once per logical field instead of retaining independent aliases
- preserve nearest-class precedence, including an explicit `None` that clears an inherited value declared through another alias
- added regression coverage for inherited core metadata, leftmost-base precedence, and cross-alias clearing

Validation: focused recipe/contract/seed-skill suite (80 passed), catalog freshness check, `git diff --check`, `./runtest.sh -s`, and pre-push agent-skill lint all passed.

**Comment by chesterxgchen** [comment]
Addressed review #168 in `e297b4f94`:

- introduced a distinct sentinel for statically unresolvable expressions
- unresolved nearest-class metadata blocks inherited aliases but is omitted so core inference/defaults still run
- explicit literal `None` remains authoritative and suppresses inference
- added regression coverage for dynamic algorithm/aggregation/state-exchange/privacy expressions and the corresponding explicit-`None` cases

Validation: focused recipe/contract/seed-skill suite (81 passed), catalog freshness check, `git diff --check`, `./runtest.sh -s`, and pre-push agent-skill lint all passed.
